core/internal/runconfig/runconfig.go

```diff
@@ -388,3 +301,51 @@ func getOrMakeSubtree(
 
 	return tree, nil
 }
+
+// Returns a deep copy of the given tree.
+func deepCopy(tree RunConfigDict) (*RunConfigDict, error) {
+	clone := make(RunConfigDict)
+	for key, value := range tree {
+		switch value := value.(type) {
+		case RunConfigDict:
+			innerClone, err := deepCopy(value)
+			if err != nil {
+				return nil, err
+			}
+			clone[key] = *innerClone
+		case []interface{}:
+			innerClone, err := deepCopyList(value)
+			if err != nil {
+				return nil, err
+			}
+			clone[key] = innerClone
```

Comment:
Hmm... can we do without deep-copying lists? There could be other mutable values in the tree that we only shallow-copy, so either way we must require that values in the tree don't get mutated.

Reply:
We will never modify list elements in our filter code but I figured why not be extra safe. Can other values get mutated? I think the leaves of the tree will always be primitives if we are using the run config and unmarshalled json/yaml

Reply:
Oh I thought that []interface{} would match lists of primitives, nvm

Reply:
I updated this so that it will deep copy slices of primitives. let me know what you think!

Reply:
There are other primitives in Go, which is why I don't think this is the right direction. For instance uint8, uint16, uint32, uint64, the signed versions, similar things for floats, and complex numbers.

I think it's better not to have special handling for slices at all. As a rule, Go code should never mutate a slice it got from elsewhere.

Reply:
Alright, I'll rip it out.